Bug in Z3 encoding of pattern-matching

When running the command : 

     dune exec compiler/catala.exe -- Proof examples/allocations_logement/allocations_logement.catala_fr -s ÉligibilitéAidePersonnelleLogement 

on the branch [`allocations_logement`](https://github.com/CatalaLang/catala/tree/allocations_logement), we get the following error message:

```
catala: internal error, uncaught exception:
        Z3.Error("Sort mismatch at argument #2 for function (declare-fun => (Bool Bool) Bool) supplied sort is Real")
        Raised by primitive operation at Z3.Boolean.mk_implies in file "../src/api/ml/z3.ml" (inlined), line 573, characters 19-38
        Called from Verification__Z3backend.translate_expr.(fun) in file "compiler/verification/z3backend.real.ml", line 732, characters 12-50
```

This error comes from the encoding in Z3 of the following expression:

```
[ERROR] Encoding error!                

   --> examples/allocations_logement/code_construction_legislatif.catala_fr
    | 
166 |     (selon ménage.logement.propriétaire sous forme
    |      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
167 |       -- DemandeurOuConjointOuParentOuViaPartsSociétés de parts: parts
    |       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
168 |       -- ParentOuAutre.Autre: 0,0) <. seuil_l822_3_parts_propriété et
    |       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    + Code de la construction et de l'habitation
    +-+ Partie législative
      +-+ Livre VIII : Aides personnelles au logement
        +-+ Titre II : Dispositions communes aux aides personnelles au logement
          +-+ Chapitre II : Conditions générales d'attribution
            +-+ Article L822-3
```

More precisely, the arms of this match are encoded as follow:

```
[DEBUG] Arm:
λ (parts_2327: decimal) → parts_2327
Z3 Arm:
(|DemandeurOuConjointOuParentOuViaPartsSociétés!0|
  (|propriétaire| (logement |ménage_979|)))
[DEBUG] Arm:
λ (__2328: unit) → 0.
Z3 Arm:
0.0
```

It seems like the encoding of the first arm, which is the identity function, is broken since it speaks about totally different things in Z3. Maybe a context problem ? It's likely that the bug comes from

https://github.com/CatalaLang/catala/blob/6ca133003e6a48a5e5d69cde7025c8aa2fe4388f/compiler/verification/z3backend.real.ml#L632-L654

--- a/compiler/verification/z3backend.real.ml
+++ b/compiler/verification/z3backend.real.ml
@@ -626,7 +626,7 @@ let rec translate_op (ctx : context) (op : operator) (args : 'm expr list) :
 
 (** [translate_expr] translate the expression [vc] to its corresponding Z3
     expression **)
-and translate_expr (ctx : context) (vc : 'm expr) : context * Expr.expr =
+and translate_expr (ctx : context) (vc : typed expr) : context * Expr.expr =
   let translate_match_arm
       (head : Expr.expr)
       (ctx : context)
@@ -817,7 +817,7 @@ module Backend = struct
 
   let is_model_empty (m : model) : bool = List.length (Z3.Model.get_decls m) = 0
 
-  let translate_expr (ctx : backend_context) (e : 'm expr) =
+  let translate_expr (ctx : backend_context) (e : typed expr) =
     translate_expr ctx e
 
   let init_backend () =

--- a/compiler/verification/z3backend.real.ml
+++ b/compiler/verification/z3backend.real.ml
@@ -705,6 +705,18 @@ and translate_expr (ctx : context) (vc : 'm expr) : context * Expr.expr =
     let ctr = List.nth ctrs idx in
     ctx, Expr.mk_app ctx.ctx_z3 ctr [z3_arg]
   | EMatch (arg, arms, enum) ->
+    (* We will encode a match as a new variable, tmp_v, and add to the hypotheses
+       that this variable is equal to the conjunction of all `A? arg ==> tmp_v == body`,
+       where `A? arg ==> body` is an arm of the match *)
+
+    (* We use the Var module to ensure that all names for temporary variables will be
+       fresh, and thus will not clash in Z3 *)
+    let fresh_v = Var.make "z3!match_tmp" in
+    let name = unique_name fresh_v in
+    let match_ty = failwith "GET TYPE" in
+    let ctx, z3_ty = translate_typ ctx (Marked.unmark match_ty) in
+    let z3_var = Expr.mk_const_s ctx.ctx_z3 name z3_ty in
+
     let ctx, z3_enum = find_or_create_enum ctx enum in
     let ctx, z3_arg = translate_expr ctx arg in
     let _ctx, z3_arms =
@@ -716,13 +728,18 @@ and translate_expr (ctx : context) (vc : 'm expr) : context * Expr.expr =
     let z3_arms =
       List.map2
         (fun r arm ->
-          (* Encodes A? arg ==> body *)
+          (* Encodes A? arg ==> z3_var = body *)
           let is_r = Expr.mk_app ctx.ctx_z3 r [z3_arg] in
-          Boolean.mk_implies ctx.ctx_z3 is_r arm)
+          let eq = Boolean.mk_eq ctx.ctx_z3 z3_var arm in
+          Boolean.mk_implies ctx.ctx_z3 is_r eq)
         (Datatype.get_recognizers z3_enum)
         z3_arms
     in
-    ctx, Boolean.mk_and ctx.ctx_z3 z3_arms
+
+    (* Add the definition of z3_var to the hypotheses *)
+    let ctx = add_z3constraint (Boolean.mk_and ctx.ctx_z3 z3_arms) ctx in
+    ctx, z3_var
+
   | EArray _ -> failwith "[Z3 encoding] EArray unsupported"
   | ELit l -> ctx, translate_lit ctx l
   | EAbs _ -> failwith "[Z3 encoding] EAbs unsupported"

--- a/compiler/verification/z3backend.real.ml
+++ b/compiler/verification/z3backend.real.ml
@@ -713,7 +713,7 @@ and translate_expr (ctx : context) (vc : typed expr) : context * Expr.expr =
        fresh, and thus will not clash in Z3 *)
     let fresh_v = Var.make "z3!match_tmp" in
     let name = unique_name fresh_v in
-    let match_ty = failwith "GET TYPE" in
+    let Typed {ty = match_ty; _ } = Marked.get_mark vc in
     let ctx, z3_ty = translate_typ ctx (Marked.unmark match_ty) in
     let z3_var = Expr.mk_const_s ctx.ctx_z3 name z3_ty in
 

--- a/compiler/verification/z3backend.real.ml
+++ b/compiler/verification/z3backend.real.ml
@@ -705,15 +705,15 @@ and translate_expr (ctx : context) (vc : typed expr) : context * Expr.expr =
     let ctr = List.nth ctrs idx in
     ctx, Expr.mk_app ctx.ctx_z3 ctr [z3_arg]
   | EMatch (arg, arms, enum) ->
-    (* We will encode a match as a new variable, tmp_v, and add to the hypotheses
-       that this variable is equal to the conjunction of all `A? arg ==> tmp_v == body`,
-       where `A? arg ==> body` is an arm of the match *)
+    (* We will encode a match as a new variable, tmp_v, and add to the
+       hypotheses that this variable is equal to the conjunction of all `A? arg
+       ==> tmp_v == body`, where `A? arg ==> body` is an arm of the match *)
 
-    (* We use the Var module to ensure that all names for temporary variables will be
-       fresh, and thus will not clash in Z3 *)
+    (* We use the Var module to ensure that all names for temporary variables
+       will be fresh, and thus will not clash in Z3 *)
     let fresh_v = Var.make "z3!match_tmp" in
     let name = unique_name fresh_v in
-    let Typed {ty = match_ty; _ } = Marked.get_mark vc in
+    let (Typed { ty = match_ty; _ }) = Marked.get_mark vc in
     let ctx, z3_ty = translate_typ ctx (Marked.unmark match_ty) in
     let z3_var = Expr.mk_const_s ctx.ctx_z3 name z3_ty in
 
@@ -739,7 +739,6 @@ and translate_expr (ctx : context) (vc : typed expr) : context * Expr.expr =
     (* Add the definition of z3_var to the hypotheses *)
     let ctx = add_z3constraint (Boolean.mk_and ctx.ctx_z3 z3_arms) ctx in
     ctx, z3_var
-
   | EArray _ -> failwith "[Z3 encoding] EArray unsupported"
   | ELit l -> ctx, translate_lit ctx l
   | EAbs _ -> failwith "[Z3 encoding] EAbs unsupported"

